demos/dev_mode_key_provisioning/src/aws_key_provisioning_demo.c

```diff
@@ -0,0 +1,613 @@
+/* Standard includes. */
+#include "string.h"
+#include "stdio.h"
+#include <stdint.h>
+
+/* FreeRTOS includes. */
+#include "FreeRTOS.h"
+#include "task.h"
+#include "message_buffer.h"
+
+/* Credentials includes. */
+#include "aws_clientcredential.h"
+
+/* Demo includes. */
+#include "aws_demo_config.h"
+#include "aws_key_provisioning_demo.h"
+
+/* Crypto includes. */
+#include "iot_pkcs11.h"
+#include "iot_pkcs11_config.h"
+#include "aws_dev_mode_key_provisioning.h"
+#include "iot_tls.h"
+
+/* mbedTLS includes. */
+#include "mbedtls/sha256.h"
+#include "mbedtls/pk.h"
+#include "mbedtls/pk_internal.h"
+#include "mbedtls/oid.h"
+#include "mbedtls/entropy.h"
+#include "mbedtls/ctr_drbg.h"
+#include "mbedtls/x509.h"
+#include "mbedtls/x509_csr.h"
+#include "mbedtls/x509_crt.h"
+
+#ifndef AWS_CLIENT_CREDENTIAL_KEYS_H
+    #define AWS_CLIENT_CREDENTIAL_KEYS_H
+
+#endif /* AWS_CLIENT_CREDENTIAL_KEYS_H */
+
+/**
+ * @brief Gives users the option to generate keys externally or internally
+ * 0 if keys are to be imported
+ * 1 if keys are to be generated on device
+ */
+#define GENERATE_KEYS_ON_DEVICE    ( 1 )
+
+/**
+ * @brief Demo is split into two parts -- before CSR/certificate generation, and after
+ * 0 to run first part of demo
+ * 1 to run second part of demo, after user has copied device and CA certificates into aws_clientcredential_keys.h
+ * NOTE: User must change value from 0 to 1 after running scripts (generating certificates) to provision device
+ */
+#define DEMO_PART                  ( 0 )
+
+/**
+ * @brief Gives users the option to reprovision each time the demo is run
+ * 0 if no reprovisioning is required
+ * 1 if user wants to reprovision the device everytime
+ */
+/* #define REPROVISION_EACH_TIME 0 */
+
+
+#define pkcs11configLABEL_DEVICE_PRIVATE_KEY_FOR_TLS    "Device Priv TLS Key"
+#define pkcs11testLABEL_DEVICE_PRIVATE_KEY_FOR_TLS      pkcs11configLABEL_DEVICE_PRIVATE_KEY_FOR_TLS
+
+#define pkcs11configLABEL_DEVICE_PUBLIC_KEY_FOR_TLS     "Device Pub TLS Key"
+#define pkcs11testLABEL_DEVICE_PUBLIC_KEY_FOR_TLS       pkcs11configLABEL_DEVICE_PUBLIC_KEY_FOR_TLS
+
+#define SHA256_DIGEST_SIZE                              32
+#define ECDSA_SIGNATURE_SIZE                            64
```

Comment:
For these defines, we can include iot_pkcs11.h and use pkcs11SHA256_DIGEST_LENGTH, pkcs11ECDSA_P256_SIGNATURE_LENGTH, and  pkcs11RSA_2048_SIGNATURE_LENGTH respectively.

This is making me realize that I was creating some un-needed defines in the test code which I should remove :)

Reply:
Modified to use iot_pkcs11.h defines instead and removed a few I didn't end up using. Thanks!